lib/differ.dart

```diff
@@ -0,0 +1,143 @@
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:json_diff/json_diff.dart';
+
+typedef Json = Map<String, dynamic>;
+
+Object diff(Object json1, Object json2) {
+  return JsonDiffer.fromJson(json1, json2).diff().toJsonString();
+}
+
+extension DiffNodeSerialization on DiffNode {
+  T apply<T>(T child) {
+    Object applied;
+    if (child is List) {
+      applied = _applyToList(child);
+    } else if (child is Map<String, dynamic>) {
+      applied = _applyToMap(child);
+    } else {
+      throw ArgumentError();
+    }
+    return applied as T;
+  }
+
+  List _applyToList(List original) {
+    final other = List.from(original);
+    for (final entry in added.entries) {
+      final key = toInt(entry.key);
+      if (key >= other.length) {
+        other.add(entry.value);
+      } else {
+        other[key] = entry.value;
+      }
+    }
+    for (final entry in changed.entries) {
+      final key = toInt(entry.key);
+      other[key] = entry.value.last;
+    }
+    for (final entry in removed.entries) {
+      final key = toInt(entry.key);
+      other.removeAt(key);
+    }
+    for (final entry in moved.entries) {
+      other.swap(entry.key, entry.value);
+    }
+    for (final entry in node.entries) {
+      final key = toInt(entry.key);
+      final child = original[key];
+      other[key] = entry.value.apply(child);
+    }
+    return other;
+  }
+
+  Map<String, dynamic> _applyToMap(Map<String, dynamic> original) {
+    final other = Map<String, dynamic>.from(original);
+    for (final entry in added.entries) {
+      final key = entry.key as String;
+      other[key] = entry.value;
+    }
+    for (final entry in changed.entries) {
+      final key = entry.key.toString();
+      other[key] = entry.value.last;
+    }
+    for (final entry in removed.entries) {
+      final key = entry.key as String;
+      other.remove(key);
+    }
+    for (final entry in node.entries) {
+      final key = entry.key as String;
+      final child = original[key];
+      other[key] = entry.value.apply(child);
+    }
+    return other;
+  }
+
+  DiffNode operator +(DiffNode other) {
```

Comment:
I don't see any guidance around operator overloading in the effective dart doc, but I tend to avoid it for non-numeric types.

Reply:
I love operator overloading for building DSLs ;) I will have to think about this..